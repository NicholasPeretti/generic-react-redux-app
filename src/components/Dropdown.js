import React from 'react'
import PropTypes from 'prop-types'
import Text from './Text'
import Style from '../style'
import styled from 'styled-components'

const DropdownContainer = styled.div`
  min-width: 128px;
  height: 32px;
`

const DropdownButton = styled.button`
  border: 2px solid ${Style.colors.gray};
  border-radius: 3px;
  background: white;
  height: 32px;
  min-width: 128px;
  width: 100%;
  text-align: left;
  cursor: pointer;
`

const DropdownButtonArrowContainer = styled.span`
  float: right;
  color: ${Style.colors.gray}
`

const DropdownListContainer = styled.div`
  width:100%;
`

const DropdownList = styled.ul`
  display: flex;
  flex-direction: column;
  list-style-type: none;
  border-radius: 3px;
  margin: 8px 0px;
  padding: 0px;
  min-width: 128px;

  li {
    cursor: pointer;
    background: white;
    padding: 8px;
    border: 2px solid ${Style.colors.gray};
    border-bottom: 0px;
    z-index: 2;

    &:hover {
      color: white;
      background: ${Style.colors.accent};
    }
  }

  li:last-child {
    border-radius: 0px 0px 3px 3px;
    border-bottom: 2px solid ${Style.colors.gray};
  }

  li:first-child {
    border-radius: 3px 3px 0px 0px;
  }
`

const DropdownItem = ({ item, onClick }) => (
  <li onClick={onClick} value={item}>
    <Text>{item}</Text>
  </li>
)

class Dropdown extends React.Component {
  state = {
    showList: false,
    value: null
  }

  static getDerivedStateFromProps (props, state) {
    const { items } = props

    if (state.value) return null

    return Object.assign({}, state, {
      value: items[0]
    })
  }

  toggleList () {
    this.setState({
      showList: !this.state.showList
    })
  }

  onSelectItem (e) {
    const value = e.target.closest('li').getAttribute('value')
    this.setState({ value, showList: false }, () => {
      this.triggerOnChange()
    })
  }

  triggerOnChange () {
    if (this.props.onChange) {
      this.props.onChange(
        this.state.value
      )
    }
  }

  render() {
    const { items, style } = this.props

    return (
      <DropdownContainer style={style}>
        <DropdownButton onClick={this.toggleList.bind(this)}>
          <Text>
            {this.state.value}
          </Text>
          <DropdownButtonArrowContainer>
            {this.state.showList ? <span>&#9650;</span> : <span>&#9660;</span>}
          </DropdownButtonArrowContainer>
        </DropdownButton>
        {this.state.showList && (
          <DropdownListContainer>
            <DropdownList>
              {items.map((item, key) =>
                <DropdownItem
                  item={item}
                  key={key}
                  onClick={this.onSelectItem.bind(this)}
                />
              )}
            </DropdownList>
          </DropdownListContainer>
        )}
      </DropdownContainer>
    )
  }
}

Dropdown.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func
}

export default Dropdown
