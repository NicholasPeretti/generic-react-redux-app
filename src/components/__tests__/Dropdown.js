import React from 'react'
import { mount } from 'enzyme'
import Dropdown from '../Dropdown'
import PresentationalTest from '../../utils/tests/presentational'

const props = {
  items: [
    'opt1',
    'opt2',
    'opt3'
  ]
}

const getWrapper = (onChange) => (
  mount(<Dropdown {...props} onChange={onChange}/>)
)

describe('Dropdown', () => {
  PresentationalTest(Dropdown, props)

  test('Should match snapshot', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should match snapshot when open', () => {
    const wrapper = getWrapper()
    wrapper.find('button').simulate('click')
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('Dropdown behaviour', () => {
    test('Should show the button', () => {
      const wrapper = getWrapper()
      expect(wrapper.exists('button')).toEqual(true)
    })

    test('Should show the list when clicked', () => {
      const wrapper = getWrapper()
      wrapper.find('button').simulate('click')
      expect(wrapper.exists('ul')).toBe(true)
    })

    test('Should hide the list when an option is clicked', () => {
      const wrapper = getWrapper()
      wrapper.find('button').simulate('click')
      wrapper.find('li').last().simulate('click')
      expect(wrapper.exists('ul')).toBe(false)
    })

    test('Should call onChange when an option is clicked', () => {
      const onChange = jest.fn()
      const wrapper = getWrapper(onChange)

      wrapper.find('button').simulate('click')
      wrapper.find('li').last().simulate('click')

      expect(onChange).toHaveBeenCalled()
    })

    test('Should set the right value after selection', () => {
      const wrapper = getWrapper()
      const { items } = props

      wrapper.find('button').simulate('click')
      wrapper.find('li').last().simulate('click')

      expect(wrapper.state().value).toEqual(
        items[items.length - 1]
      )
    })
  })
})
