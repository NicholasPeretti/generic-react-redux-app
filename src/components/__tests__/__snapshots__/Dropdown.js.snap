// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Should match snapshot 1`] = `"<div class=\\"sc-bwzfXH gYNwqV\\"><button class=\\"sc-htpNat kYidVU\\"><span class=\\"sc-bdVaJa bkmfbd\\">opt1</span><span class=\\"sc-bxivhb kAOnLj\\"><span>▼</span></span></button></div>"`;

exports[`Dropdown Should match snapshot when open 1`] = `"<div class=\\"sc-bwzfXH gYNwqV\\"><button class=\\"sc-htpNat kYidVU\\"><span class=\\"sc-bdVaJa bkmfbd\\">opt1</span><span class=\\"sc-bxivhb kAOnLj\\"><span>▲</span></span></button><div class=\\"sc-ifAKCX jBGnp\\"><ul class=\\"sc-EHOje huREYR\\"><li value=\\"opt1\\"><span class=\\"sc-bdVaJa bkmfbd\\">opt1</span></li><li value=\\"opt2\\"><span class=\\"sc-bdVaJa bkmfbd\\">opt2</span></li><li value=\\"opt3\\"><span class=\\"sc-bdVaJa bkmfbd\\">opt3</span></li></ul></div></div>"`;
